### DISCUSSION TITLE
Do you plan to add suport for ES Modules?

### DISCUSSION THREAD

**Discussion by gr2m** [johnlindquist/kit#242] [Q&A]
When trying to require a esm package, e.g.

```js
const { Octokit } = await npm("@octoherd/octokit");
```

I get the following error

```
import { Octokit as OctokitCore } from "@octokit/core";
^^^^^^

SyntaxError: Cannot use import statement outside a module
    at Object.compileFunction (node:vm:355:18)
    at wrapSafe (node:internal/modules/cjs/loader:1038:15)
    at Module._compile (node:internal/modules/cjs/loader:1072:27)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1137:10)
    at Module.load (node:internal/modules/cjs/loader:988:32)
    at Function.Module._load (node:internal/modules/cjs/loader:828:14)
    at Module.require (node:internal/modules/cjs/loader:1012:19)
    at require (node:internal/modules/cjs/helpers:93:18)
    at global.npm (/Users/gregor/.kit/preload/mac-app.cjs:732:12)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
```

**Comment by johnlindquist** [comment]
Yeah, I was trying to keep `npm` as a just-in-time/lazy installer, but that only works with `require`. I'll need to pre-parse the script from the app to support esm. I haven't really thought about the "flow" of it yet.

For now I'd just use 

```js
const { Octokit } = await import("@octoherd/octokit");
```

You can install packages with `Help->Manage npm packages->install` or `~/.kit/node/bin/npm i @octokit/octokit --prefix ~/.kenv`

**Reply by gr2m** [reply]
> Yeah, I was trying to keep `npm` as a just-in-time/lazy installer, but that only works with `require`

Maybe worth using `deno` instead of Node?

I think the `await npm()` method should be able to do anything you want, such as

- prompting the user whether to install
- install the dependency
- use the async `import()` method 

I don't see why this shouldn't work with ESM?

Script Kit is a new project, I'd go all in on ESM, it really is a nice developer experience

**Reply by johnlindquist** [reply]
Deno currently doesn't support preloading scripts. I'm keeping an eye on it. And we'll see what the future brings.

Esm requires the package to already be installed in package.json when Node starts. Cjs require can get around that restriction and "lazy install" then parse the package.json. It's possible I'm missing something, I've haven't revisited that code in a while.

I'll most likely switch npm to using Esm regardless, I'll need to parse scripts before they run to check for missing packages which is "fine", just misses out on that "just in time" feel.
